### PR TITLE
Model outputs returns unblended RGB (black bg) for random bg

### PR DESCRIPTION
**Pull request by jkulhanek**

### CODE REVIEW

**Comment by jkulhanek** [comment]
Hi, I think I have found a solution to simplify the impl a bit and to return nice images for metrics computation (as blended with black colour). Let me know what you think.

File: nerfstudio/model_components/renderers.py [hunk 1]
```diff
@@ -124,65 +128,68 @@ def get_background_color(
...
             background_color: Background color.
 
         Returns:
             Blended RGB.
         """
-        if len(opacity.shape) < len(rgb.shape):
-            opacity = opacity.unsqueeze(-1)
+        if background_color is None:
+            background_color_name = self.background_color
+            if background_color_name == "last_sample" or background_color == "random":
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py]: should be `background_color_name == "random"`

File: nerfstudio/model_components/renderers.py [hunk 2]
```diff
@@ -124,65 +128,68 @@ def get_background_color(
...
+                background_color_name = "black"
+            background_color = self.get_background_color(
+                rgb=image.unsqueeze(-1), background_color=background_color_name
+            )
+        if image.size(-1) < 4:
+            return image
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py:163]: Better to move this before `if background_color is None:` to avoid unnecessary computations

File: nerfstudio/model_components/renderers.py [hunk 3]
```diff
@@ -124,65 +128,68 @@ def get_background_color(
...
-            )
-            gt_rgb = self.blend_background(
-                rgb=gt_image[..., :3], opacity=gt_image[..., 3], background_color=background_color
-            )
-        else:
-            gt_rgb = gt_image[..., :3]
 
-        return predicted_rgb, gt_rgb
+        if self.background_color == "random":
+            background_color = torch.rand_like(pred_image)
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py:194]: Why move this here rather than leaving it in `get_background_color`? The "last_sample" color also isn't constant

File: nerfstudio/model_components/renderers.py [hunk 4]
```diff
@@ -124,65 +128,68 @@ def get_background_color(
...
 
         Returns:
             Blended RGB.
         """
-        if len(opacity.shape) < len(rgb.shape):
-            opacity = opacity.unsqueeze(-1)
+        if background_color is None:
+            background_color_name = self.background_color
+            if background_color_name == "last_sample" or background_color == "random":
+                background_color_name = "black"
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py]: I find this slightly confusing based on the function name. What about adding a function `blend_background_for_metric_computation`, which does this logic before calling `blend_background`? `blend_background` could then make `background_color` a required argument.

File: nerfstudio/model_components/renderers.py [hunk 5]
```diff
@@ -101,9 +102,12 @@ def combine_rgb(
             comp_rgb = torch.sum(weights * rgb, dim=-2)
             accumulated_weight = torch.sum(weights, dim=-2)
 
+        if background_color == "random":
+            # If background color is random, the predicted color is blended with black (0, 0, 0).
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py]: Using "added" instead of "blended" would be more accurate

File: nerfstudio/models/nerfacto.py [hunk 6]
```diff
@@ -345,14 +333,13 @@ def get_metrics_dict(self, outputs, batch):
     def get_loss_dict(self, outputs, batch, metrics_dict=None):
         loss_dict = {}
         image = batch["image"].to(self.device)
-        blended_rgb = (
-            self.renderer_rgb.blend_background(
-                rgb=image[..., :3], opacity=image[..., 3], background_color=outputs["background_color"]
-            )
-            if image.shape[-1] == 4 and "background_color" in outputs
-            else image[..., :3]
+        pred_rgb, gt_rgb = self.renderer_rgb.blend_background_for_loss_computation(
```
Comment (nepfaff) [inline nerfstudio/models/nerfacto.py:336]: Wouldn't this add the background color twice if the background color isn't random? First time in `combine_rgb`.

Comment (nepfaff) [reply nerfstudio/models/nerfacto.py:336]: What about removing it from `combine_rgb` and making sure that we do this through one of the other methods if required? It is more of a trick rather than being part of volume rendering anyways.

**Review by nepfaff — COMMENTED** [review]
Looks great! I think this refactoring makes it much better, and we should definitely adapt those changes. Thanks!

File: nerfstudio/model_components/renderers.py (hunk 3 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py:194]: I will leave it there. I was a bit afraid people would use get_background_color for their own stuff without being aware of the risks.

Comment (nepfaff) [reply nerfstudio/model_components/renderers.py:194]: I see what you mean. Maybe `get` is a bad name. What about renaming it to `compute`, making it a bit clearer that the output might be different everytime that it is called?

File: nerfstudio/model_components/renderers.py (hunk 2 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py:163]: Sure, sorry

File: nerfstudio/model_components/renderers.py (hunk 4 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py]: I believe the logic here is fine. It is not well defined for random/last_sample anyway. I feel like this is better than adding one more method?

File: nerfstudio/models/nerfacto.py (hunk 6 above)
Comment (jkulhanek) [reply nerfstudio/models/nerfacto.py:336]: Yes, you are right! It was adding the background color twice.

**Comment by nepfaff** [comment]
Do you know if there is a way for me to checkout this commit for local testing?

File: nerfstudio/models/nerfacto.py (hunk 6 above)
Comment (jkulhanek) [reply nerfstudio/models/nerfacto.py:336]: Do you mean moving it to `forward`? We would need to introduce the "add_background" method (as blending cannot be used for the case).

File: nerfstudio/model_components/renderers.py (hunk 4 above)
Comment (nepfaff) [reply nerfstudio/model_components/renderers.py]: sure :)

This is very much personal preference, so I'm happy to leave it like this

File: nerfstudio/model_components/renderers.py [hunk 7]
```diff
@@ -130,59 +134,60 @@ def get_background_color(
...
-                rgb=gt_image[..., :3], opacity=gt_image[..., 3], background_color=background_color
-            )
-        else:
-            gt_rgb = gt_image[..., :3]
 
-        return predicted_rgb, gt_rgb
+        background_color = self.get_background_color(
+            rgb=pred_image.unsqueeze(-1), background_color=self.background_color
+        ).to(pred_image.device)
+        if self.background_color == "random":
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py]: Looks like this if statement is still there from the last change

**Comment by jkulhanek** [comment]
Yes, you can checkout the `nepfaff/alpha_carving_v2` from `nerfstudio-project` origin. Have you added the nerfstudio origin to your local fork? If not, you can do the following:
```
git remote add origin nerfstudio git@github.com:nerfstudio-project/nerfstudio.git
git fetch nerfstudio
git checkout nepfaff/alpha_carving_v2
```

File: nerfstudio/model_components/renderers.py (hunk 7 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py]: Now I think it is correct, no? The predicted RGB is only "blended" if self.background_color == random. The GT is always blended (as it can be transparent here).

**Comment by jkulhanek** [comment]
(Not sure if the commands I send you will work correctly)

File: nerfstudio/models/nerfacto.py (hunk 6 above)
Comment (nepfaff) [reply nerfstudio/models/nerfacto.py:336]: I was thinking in terms of having one function, `blend_background_for_metric_computation` and one `blend_background_for_loss_computation`. One of them would always be called and would do the background color adding. Having it in `forward` wouldn't prevent the adding twice problem.

If we don't want a `blend_background_for_metric_computation` method, we could add a flag to `forward` to determine whether to add or not.

File: nerfstudio/model_components/renderers.py (hunk 5 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py]: Updated the comment.

**Comment by nepfaff** [comment]
It worked. Thanks!

File: nerfstudio/model_components/renderers.py (hunk 7 above)
Comment (nepfaff) [reply nerfstudio/model_components/renderers.py]: Ah, I see what you mean. So this is resolving the double blending issue by blending "random" here and everything else as done before?

Comment (nepfaff) [reply nerfstudio/model_components/renderers.py]: In this case, I think it is fine like this and no need to do anything for https://github.com/nepfaff/nerfstudio/pull/1#discussion_r1265175381

**Comment by jkulhanek** [comment]
I have added other methods.

File: nerfstudio/models/nerfacto.py (hunk 6 above)
Comment (jkulhanek) [reply nerfstudio/models/nerfacto.py:336]: Is this still an issue or was it resolved?

File: nerfstudio/model_components/renderers.py (hunk 3 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py:194]: Did you introduce this function or was it in nerfstudio before?

Comment (nepfaff) [reply nerfstudio/model_components/renderers.py:194]: I introduced it. It used to be part of `combine_rgb`

File: nerfstudio/models/nerfacto.py (hunk 6 above)
Comment (nepfaff) [reply nerfstudio/models/nerfacto.py:336]: Resolved with https://github.com/nepfaff/nerfstudio/pull/1#discussion_r1265211859

File: nerfstudio/model_components/renderers.py (hunk 3 above)
Comment (jkulhanek) [reply nerfstudio/model_components/renderers.py:194]: Great, then we are free to define it as we see fit. I will do some refactoring.

**Comment by jkulhanek** [comment]
@nepfaff Can you please take a look now?

**Comment by nepfaff** [comment]
Done some testing, and the results aren't great yet. The implicit blending with black is causing issues for export.
Same problem as currently exists in the previous version: https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1263880289

The following are after 4k iterations.

### Current code
current03

### Modification

I added
```python
if not self.training:
            background_color = "last_sample"
```
to `forward`.

modified04

**Comment by jkulhanek** [comment]
> Done some testing, and the results aren't great yet. The implicit blending with black is causing issues for export. Same problem as currently exists in the previous version: [nerfstudio-project#2165 (comment)](https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1263880289)
> 
> The following are after 4k iterations.
> 
> ### Current code
> current03
> 
> ### Modification
> I added
> 
> ```python
> if not self.training:
>             background_color = "last_sample"
> ```
> 
> to `forward`.
> 
> modified04

Seems like a hack a bit. There is no reason for the last_sample to return good colours, is there? We can export the transparent PC (or threshold it at some alpha). What do you think?

File: nerfstudio/model_components/renderers.py [hunk 8]
```diff
@@ -101,14 +102,23 @@ def combine_rgb(
...
+        assert isinstance(background_color, torch.Tensor)
+        comp_rgb = comp_rgb + background_color * (1.0 - accumulated_weight)
         return comp_rgb
 
     @classmethod
     def get_background_color(
-        cls, rgb: Float[Tensor, "*bs num_samples 3"], background_color: BackgroundColor
+        cls, background_color: BackgroundColor, shape: Tuple[int, ...], device: torch.device
     ) -> Union[Float[Tensor, "3"], Float[Tensor, "*bs 3"]]:
         """Returns the RGB background color for a specified background color.
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py:123]: I'd add a comment that it can't be called for "last_sample" and "random".

File: nerfstudio/model_components/renderers.py [hunk 9]
```diff
@@ -119,70 +129,69 @@ def get_background_color(
...
-                rgb=gt_image[..., :3], opacity=gt_image[..., 3], background_color=background_color
-            )
-        else:
-            gt_rgb = gt_image[..., :3]
-
-        return predicted_rgb, gt_rgb
+        background_color = self.background_color
+        if self.background_color == "last_sample":
+            background_color = "black"  # No background blending for GT
+        elif self.background_color == "random":
```
Comment (nepfaff) [inline nerfstudio/model_components/renderers.py]: It would be slightly cleaner to use `background_color` instead of `self.background_color`.

**Comment by jkulhanek** [comment]
> > Done some testing, and the results aren't great yet. The implicit blending with black is causing issues for export. Same problem as currently exists in the previous version: [nerfstudio-project#2165 (comment)](https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1263880289)
> > The following are after 4k iterations.
> > ### Current code
> > current03
> > ### Modification
> > I added
> > ```python
> > if not self.training:
> >             background_color = "last_sample"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to `forward`.
> > modified04
> 
> Seems like a hack a bit. There is no reason for the last_sample to return good colours, is there? We can export the transparent PC (or threshold it at some alpha). What do you think?

Can you please (for testing) just try dividing the RGB by accumulation in exporter_utils.py:143?

**Comment by nepfaff** [comment]
> > Done some testing, and the results aren't great yet. The implicit blending with black is causing issues for export. Same problem as currently exists in the previous version: [nerfstudio-project#2165 (comment)](https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1263880289)
> > The following are after 4k iterations.
> > ### Current code
> > current03
> > ### Modification
> > I added
> > ```python
> > if not self.training:
> >             background_color = "last_sample"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to `forward`.
> > modified04
> 
> Seems like a hack a bit. There is no reason for the last_sample to return good colours, is there? We can export the transparent PC (or threshold it at some alpha). What do you think?

This is definitely a hack, and I only used it for illustrative purposes. I chose this as its the default background color for nerfacto

**Comment by jkulhanek** [comment]
I don't think we should export a transparent point cloud. Does normalizing by acc (in pc generation) give the correct colours?

**Comment by nepfaff** [comment]
> I don't think we should export a transparent point cloud. Does normalizing by acc (in pc generation) give the correct colours?

It is not visible enough with a pcd output. I'm currently testing it for the TSDF export.

**Comment by nepfaff** [comment]
> > I don't think we should export a transparent point cloud. Does normalizing by acc (in pc generation) give the correct colours?
> 
> It is not visible enough with a pcd output. I'm currently testing it for the TSDF export.

Changing line 408 in `texture_utils.py` to the following works.
```python
texture_image = (outputs["rgb"] / outputs["accumulation"]).cpu().numpy()
```

**Comment by nepfaff** [comment]
> > > I don't think we should export a transparent point cloud. Does normalizing by acc (in pc generation) give the correct colours?
> > 
> > 
> > It is not visible enough with a pcd output. I'm currently testing it for the TSDF export.
> 
> Changing line 408 in `texture_utils.py` to the following works.
> 
> ```python
> texture_image = (outputs["rgb"] / outputs["accumulation"]).cpu().numpy()
> ```

Do you think changing this everywhere will have side effects?

**Comment by jkulhanek** [comment]
> changing

Yes, we have to do this only if background_color=random. Furthermore, we should do some thresholding based on the accumulation, e.g. discard if acc<0.5.

Does it help in the current case?

**Comment by nepfaff** [comment]
Yes, it fixes the issue completely. Visually, the effect is the same as when using the hack from here https://github.com/nepfaff/nerfstudio/pull/1#issuecomment-1638024731

**Comment by nepfaff** [comment]
What do you mean by discard?

**Comment by jkulhanek** [comment]
I mean to delete the points from the PC. I will merge this PR and we will continue in the main PR, is it ok?

**Comment by nepfaff** [comment]
> I mean to delete the points from the PC. I will merge this PR and we will continue in the main PR, is it ok?

Sounds great!

**Comment by nepfaff** [comment]
Fair, for PC this would make sense
